pkg/search/proxy/store/util.go

```diff
@@ -195,9 +195,21 @@ func (w *watchMux) AddSource(watcher watch.Interface, decorator func(watch.Event
 
 // Start run the watcher
 func (w *watchMux) Start() {
+	wg := sync.WaitGroup{}
 	for _, source := range w.sources {
-		go w.startWatchSource(source.watcher, source.decorator)
+		source := source
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			w.startWatchSource(source.watcher, source.decorator)
+		}()
 	}
+
+	go func() {
+		// close result chan after all goroutines exit, avoiding data race.
```

Comment:
Just a question, why would there be data race in the previous approach?

Reply:
@ikaven1024 explained it at https://github.com/karmada-io/karmada/pull/4212#issuecomment-1805660860.